trace/adapter.go

```diff
@@ -0,0 +1,100 @@
+package trace
+
+import (
+	"context"
+	"net/http"
+
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+// GRPCServerUnaryInterceptor grpc unary server interceptor.
+func GRPCServerUnaryInterceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+	newCtx := SpanContextFrom(ctx, SpanContextFromGrpc)
+	return handler(newCtx, req)
+}
+
+type wrappedStream struct {
+	ctx context.Context
+	grpc.ServerStream
+}
+
+func (ws wrappedStream) Context() context.Context {
+	if ws.ctx != nil {
+		return ws.ctx
+	}
+	return ws.ServerStream.Context()
+}
+
+// ServerStreamInterceptor server stream.
+func GRPCServerStreamInterceptor(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+	newCtx := SpanContextFrom(ss.Context(), SpanContextFromGrpc)
+	return handler(srv, &wrappedStream{ServerStream: ss, ctx: newCtx})
+}
+
+// SpanContextGrpc gets span context from grpc request.
+func SpanContextFromGrpc(ctx context.Context) trace.SpanContext {
+	md, ok := metadata.FromIncomingContext(ctx)
```

Comment:
This method does a lot of re-allocations and string transformations on every call, and it's very slow. Not what we should use with a logger.... I changed it in the runtime to cache the values: https://github.com/dapr/dapr/blob/master/pkg/grpc/metadata/metadata.go

Perhaps consider moving that package (pkg/grpc/metadata) into dapr/kit, and use the `FromIncomingContext` method from that package.